tensorboard/plugins/custom_scalar/tf_custom_scalar_dashboard/tf-custom-scalar-margin-chart-card.html

```diff
@@ -0,0 +1,579 @@
+                title: 'Smoothed',
+                evaluate: (d, statusObject) => formatValueOrNaN(
+                    statusObject.smoothingEnabled ? d.datum.smoothed :
+                                                    d.datum.scalar),
+              },
+              {
+                title: 'Value',
+                evaluate: (d) => formatValueOrNaN(d.datum.scalar),
+              },
+              {
+                title: 'Lower Margin',
+                evaluate: (d) => formatValueOrNaN(d.datum.lower),
+              },
+              {
+                title: 'Upper Margin',
+                evaluate: (d) => formatValueOrNaN(d.datum.upper),
+              },
+              {
+                title: 'Step',
+                evaluate: (d) => vz_line_chart.stepFormatter(d.datum.step),
+              },
+              {
+                title: 'Time',
+                evaluate: (d) => vz_line_chart.timeFormatter(d.datum.wall_time),
+              },
+              {
+                title: 'Relative',
+                evaluate: (d) => vz_line_chart.relativeFormatter(
+                    vz_line_chart.relativeAccessor(d.datum, -1, d.dataset)),
+              },
+            ];
+          }
+        },
+        _tagsWithNoData: String,
+      },
+      observers: [
+        '_updateChart(_nameToDataSeries)',
+        // Clear the data series when the tag filter changes.
+        '_refreshDataSeries(_tagFilter)',
+      ],
+      reload() {
+        this.$$('tf-line-chart-data-loader').reload();
+      },
+      redraw() {
+        this.$$('tf-line-chart-data-loader').redraw();
+      },
+      _toggleExpanded(e) {
+        this.set('_expanded', !this._expanded);
+        this.redraw();
+      },
+      _toggleLogScale() {
+        this.set('_logScaleActive', !this._logScaleActive);
+      },
+      _resetDomain() {
+        const chart = this.$$('tf-line-chart-data-loader');
+        if (chart) {
+          chart.resetDomain();
+        }
+      },
+      _csvUrl(nameToDataSeries, dataSeriesName) {
+        const baseUrl = this._downloadDataUrl(nameToDataSeries, dataSeriesName);
+        return tf_backend.addParams(baseUrl, {format: 'csv'});
+      },
+      _jsonUrl(nameToDataSeries, dataSeriesName) {
+        const baseUrl = this._downloadDataUrl(nameToDataSeries, dataSeriesName);
+        return tf_backend.addParams(baseUrl, {format: 'json'});
+      },
+      _downloadDataUrl(nameToDataSeries, dataSeriesName) {
+        const dataSeries = nameToDataSeries[dataSeriesName];
+        const getVars = {
+            tag: dataSeries.getTag(),
+            run: dataSeries.getRun(),
+        };
+        return tf_backend.addParams(
+              tf_backend.getRouter().pluginRoute(
+                  'custom_scalars', '/download_data'), getVars);
+      },
+      _computeXComponentsCreationMethod(xType) {
+        return () => vz_line_chart.getXComponents(xType);
+      },
+      _createProcessDataFunction(marginChartSeries) {
```

Comment:
This function is quite long (117 lines).  Maybe it's possible to break it down a little into some subfunctions?  E.g. you could have one that takes `scalarEvents`, `lowerBounds`, and `upperBounds` and produces `dataSeries`.

Reply:
Yeah, this is much better, thanks.